File: codegen/lib/prodAtx/prodAtx.c
#include "prodAtx.h"
#include "omp.h"
#include "m2c.h"

static void b_prodAtx(const emxArray_real_T *A, const emxArray_real_T *x,
                      emxArray_real_T *b);
static void c_prodAtx(const emxArray_real_T *A, const emxArray_real_T *x,
                      emxArray_real_T *b);
static void m2c_error(void);
static void m2c_warn(void);
static void prodAtx_internal(const emxArray_real_T *A, const emxArray_real_T *x,
  emxArray_real_T *b);
static double rt_roundd(double u);
static void b_prodAtx(const emxArray_real_T *A, const emxArray_real_T *x,
                      emxArray_real_T *b)
{
  int i0;
  i0 = b->size[0] * b->size[1];
  b->size[0] = A->size[1];
  b->size[1] = x->size[1];
  emxEnsureCapacity((emxArray__common *)b, i0, (int)sizeof(double));
  prodAtx_internal(A, x, b);
}

static void c_prodAtx(const emxArray_real_T *A, const emxArray_real_T *x,
                      emxArray_real_T *b)
{
  if ((b->size[0] < A->size[1]) || (b->size[1] < x->size[1])) {
    m2c_error();
  }

  prodAtx_internal(A, x, b);
}

static void m2c_error(void)
{
  M2C_error("prodAtx:BufferTooSmal", "Buffer space for output b is too small.");
}

static void m2c_warn(void)
{
  M2C_warn("prodAtx:NestedParallel",
           "You are trying to use nested parallel regions. Solution may be incorrect.");
}

static void prodAtx_internal(const emxArray_real_T *A, const emxArray_real_T *x,
  emxArray_real_T *b)
{
  int nthreads;
  int thr_id;
  int jstart;
  double y;
  int jend;
  double b_y;
  int j;
  int chunk;
  int b_remainder;
  int k;
  int c_remainder;
  double t;
  int i;
  nthreads = omp_get_num_threads();
  if (nthreads == 1) {
    jstart = 0;
    jend = A->size[1];
  } else {
    thr_id = omp_get_thread_num();
    y = (double)A->size[1] / (double)nthreads;
    if (y < 0.0) {
      b_y = ceil(y);
    } else {
      b_y = floor(y);
    }

    chunk = (int)rt_roundd(b_y);
    b_remainder = (int)((double)A->size[1] - (double)(nthreads * chunk));
    if (b_remainder <= thr_id) {
      c_remainder = b_remainder;
    } else {
      c_remainder = thr_id;
    }

    jstart = thr_id * chunk + c_remainder;
    jend = (jstart + chunk) + (thr_id < b_remainder);
  }

  for (j = jstart; j + 1 <= jend; j++) {
    for (k = 0; k + 1 <= x->size[1]; k++) {
      t = 0.0;
      for (i = 0; i + 1 <= A->size[0]; i++) {
        t += A->data[i + A->size[0] * j] * x->data[i + x->size[0] * k];
      }

      b->data[j + b->size[0] * k] = t;
    }
  }
}

static double rt_roundd(double u)
{
  double y;
  if (fabs(u) < 4.503599627370496E+15) {
    if (u >= 0.5) {
      y = floor(u + 0.5);
    } else if (u > -0.5) {
      y = 0.0;
    } else {
      y = ceil(u - 0.5);
    }
  } else {
    y = u;
  }

  return y;
}

void emxInitArray_real_T(emxArray_real_T **pEmxArray, int numDimensions)
{
  emxInit_real_T(pEmxArray, numDimensions);
}

void prodAtx(const emxArray_real_T *A, const emxArray_real_T *x, emxArray_real_T
             *b, int nthreads)
{
  int n;
  int b_n;
  if ((b->size[0] < A->size[1]) || (b->size[1] < x->size[1])) {
    m2c_error();
  }

  n = omp_get_num_threads();
  b_n = omp_get_nested();
  if ((!(b_n != 0)) && (n > 1) && (nthreads > 1)) {

#pragma omp master

    M2C_BEGIN_REGION(/*omp master*/)

    m2c_warn();

    M2C_END_REGION(/*omp master*/)

  }

#pragma omp parallel default(shared) num_threads(nthreads)
  M2C_BEGIN_REGION(/*omp parallel*/)

  prodAtx_internal(A, x, b);

  M2C_END_REGION(/*omp parallel*/)

}

void prodAtx_initialize(void)
{
}

void prodAtx_ser(const emxArray_real_T *A, const emxArray_real_T *x,
                 emxArray_real_T *b)
{
  b_prodAtx(A, x, b);
}

void prodAtx_ser1(const emxArray_real_T *A, const emxArray_real_T *x,
                  emxArray_real_T *b)
{
  c_prodAtx(A, x, b);
}

void prodAtx_terminate(void)
{
}

File: prodAtx.m
function b = prodAtx(A, x, b, nthreads, varargin)
%prodAtx Compute b=A'*x, assuming that size(A,1)==size(x,1).
%
%      b = prodAtx(A, x [, b])
% Computes b=A'*x in serial.
%
%      b = prodAtx(A, x, b, nthreads)
% Computes b=A'*x using nthreads OpenMP threads, where nthreads is an int32.
% In this mode, there are implicit barriers in the function.
%
%      b = prodAtx(A, x, b, [])
% Computes b=A'*x locally within each OpenMP thread, assuming the
% parallel region has already been initialized. The argument b has
% been preallocated. In this mode, there is no implicit barrier.
%
%      b = prodAtx(A, x, b, nthreads, comm)
% also performs communication within the MPI communicator.
% It assumes that A is partitioned rowwise, and hence A'*x is
% a partial sum on each process. An allreduce is performed on b
% within the communicator.
%
% Note that if A is partitioned columnwise, then b must be duplicated
% on all processes, and prodAtx should be called without the communicator.
%
%      b = prodAtx(A, x, b, nthreads, comm, pgmsg [, pgsz])
% specifies an additiona message that should be summed during
% allreduce and strored in b. This is to piggy-back on allreduce
% to help aggregate small messages to optimize MPI communication.
%
% The compiled version reqiures A and x to be double precision.
% The M file works for both single and double precision.

%#codegen -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]), ...
%#codegen coder.typeof(0, [inf,inf]), int32(1)}
%#codegen prodAtx_ser -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf])}
%#codegen prodAtx_ser1 -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf])}
% %#codegen prodAtx_mpi -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), MPI_Comm}
% %#codegen prodAtx_mpip -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,1]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), MPI_Comm, coder.typeof(0, [inf,inf])}
% %#codegen prodAtx_mpip1 -args {coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), MPI_Comm, coder.typeof(0, [inf,inf]), int32(0)}

coder.inline('never');

if nargin==2;
    b = nullcopy(zeros(size(A,2),size(x,2)));
else
    if size(b,1)<size(A,2) || size(b,2)<size(x,2)
        m2c_error('prodAtx:BufferTooSmal', 'Buffer space for output b is too small.');
    end
end
ismt = nargin>3 && momp_get_num_threads>1;

if nargin>3 && ~isempty(nthreads)
    %% Declare parallel region
    if ~momp_get_nested && ismt && nthreads(1)>1
        momp_begin_master
        m2c_warn('prodAtx:NestedParallel', ...
            'You are trying to use nested parallel regions. Solution may be incorrect.');
        momp_end_master
    end
    
    momp_begin_parallel; momp_clause_default('shared');
    momp_clause_num_threads(int32(nthreads(1)));
    
    %% Compute b=A'*x
    b = prodAtx_internal(A, x, b);
    
    momp_end_parallel
else
    %% Compute b=A'*x
    b = prodAtx_internal(A, x, b);
end

if ~isempty(varargin)
    if isempty(nthreads) && ismt; momp_barrier; end
    
    % Perform MPI allreduce
    momp_begin_single
    s = int32(size(A,2)*size(x,2));
    b = allreduce(b, s, MPI_SUM, varargin{:});
    momp_end_single
end

function b = prodAtx_internal(A, x, b)
% Internal function for computing prodAtx on each thread

coder.inline('never');

nrows = int32(size(A,1)); nrhs = int32(size(x,2));

% Decompose the work manually
[jstart, jend] = get_local_chunk(size(A,2));
for j=jstart:jend
    for k=1:nrhs
        t = 0;
        for i=1:nrows
            t = t + A(i,j) * x(i,k);
        end
        b(j,k) = t;
    end
end

function test %#ok<DEFNU>
%!test
%! if ~exist(['prodAtx.' mexext], 'file')
%!    m=200; n = 20;
%! else
%!    m=10000; n = 2000;
%! end
%! tic; A = rand(m,n); x=rand(m,2);
%! fprintf(1, '\n\tGenerated random matrix in %g seconds\n', toc);
%! tic; b0 = A'*x;
%! fprintf(1, '\tComputed reference solution in %g seconds\n', toc);
%! fprintf(1, '\tTesting serial: ');
%! tic; b1 = prodAtx(A, x);
%! fprintf(1, 'Done in %g seconds\n ', toc);
%! assert(norm(b0-b1)/norm(b0)<=1.e-12);

%! b2 = zeros(n,2);
%! for nthreads=int32([1 2 4 8])
%!     if nthreads>momp_get_max_threads; break; end
%!     fprintf(1, '\tTesting %d thread(s): ', nthreads);
%!     tic; b2 = prodAtx(A, x, b2, nthreads);
%!     fprintf(1, 'Done in %g seconds\n ', toc);
%!     assert(norm(b0-b2)/norm(b0)<=1.e-12);
%! end

% %! fprintf(1, '\tTesting 2 threads with MPI call: ');
% %! tic; b2 = prodAtx(A, x, b2, int32(2), MPI_COMM_SELF);
% %! fprintf(1, 'Done in %g seconds\n', toc);
% %! assert(norm(b0-b2)/norm(b0)<=1.e-12);
% 
% %! b3 = zeros(n+1,1);
% %! fprintf(1, '\tTesting 2 threads with piggy-back: ');
% %! tic; b3 = prodAtx(A, x(:,1), b3, int32(2), MPI_COMM_WORLD, 1, int32(1));
% %! fprintf(1, 'Done in %g seconds\n', toc);
% %! assert(norm(b0(:,1)-b3(1:end-1))/norm(b0(:,1))<=1.e-12);
% %! assert(b3(end)==1);
